src/sagetasks/nextflowtower/utils.py

```diff
@@ -0,0 +1,177 @@
+
+        Optionally, you can set a default workspace for various methods with the
+        `open_workspace()` method.
+        """
+        # TODO: Validate access token
+        self.client = TowerClient(**client_args)
+        self._workspace = None
+
+    @property
+    def workspace(self):
+        """Property for raising an error if a workspace hasn't been opened yet."""
+        if self._workspace is None:
+            raise ValueError("Workspace not opened yet. Use `open_workspace()`.")
+        return self._workspace
+
+    def open_workspace(self, workspace_id):
+        """Sets the given workspace as the default for other methods."""
+        self._workspace = workspace_id
+
+    def close_workspace(self):
+        """Clear the opened (default) workspace."""
+        self._workspace = None
+
+    def init_params(self):
+        """Sets the given workspace as the default for other methods."""
+        return {"workspaceId": self.workspace}
+
+    @staticmethod
+    def extract_resource(response, key=None):
+        """Extracts the nested resource from a response dictionary.
+
+        These IDs are strings, which can be pickled unlike the resource objects.
+        """
+        if key is None and isinstance(response, dict) and len(response) == 1:
+            key = list(response)[0]
+            resource = response[key]
+        elif isinstance(response, dict) and key in response:
+            resource = response[key]
+        else:
+            resource = response
+        return resource
+
+    @staticmethod
+    def bundle_client_args(auth_token, platform="tower.nf", endpoint=None, **kwargs):
+        """Bundles the information for authenticating a Tower client."""
+        assert platform is None or platform in ENDPOINTS
+        assert platform or endpoint
+        endpoint = endpoint if endpoint else ENDPOINTS[platform]
+        client_args = dict(tower_token=auth_token, tower_api_url=endpoint, **kwargs)
+        return client_args
+
+    def get_compute_env(self, compute_env_id):
+        endpoint = f"/compute-envs/{compute_env_id}"
+        params = self.init_params()
+        response = self.client.request("GET", endpoint, params=params)
+        compute_env = self.extract_resource(response)
+        return compute_env
+
+    def get_workflow(self, workflow_id):
+        endpoint = f"/workflow/{workflow_id}"
+        params = self.init_params()
+        response = self.client.request("GET", endpoint, params=params)
+        return response
+
+    @staticmethod
+    def update_dict(base_dict, overrides):
+        dict_copy = copy(base_dict)
+        for k, v in overrides.items():
+            oldv = dict_copy.get(k, {})
+            if k not in dict_copy:
+                valid = set(dict_copy)
+                raise ValueError(f"Cannot update {k}. Not among {valid}.")
+            elif isinstance(oldv, Mapping) and isinstance(v, Mapping):
+                dict_copy[k] = TowerUtils.update_dict(oldv, v)
```

Comment:
can this be self.update_dict?

Reply:
Nope, because `update_dict()` is a static method and the method arguments don't include `self`. I could make it a regular (`self`) or class (`cls`) method, but this function doesn't need any state. Once it's used somewhere else, I'll likely migrate it to the centralized `utils` submodule.

Reply:
Ah yes, I didn't notice the static_method. Thanks